File: Ros2_bridge/src/joystick_cpp/src/cmd_vel_repeater.cpp
#include <chrono>
#include <cstdio>
#include <cstring>
#include <mutex>
#include <string>
#include <thread>

#include <termios.h>
#include <unistd.h>

#include "geometry_msgs/msg/twist.hpp"
#include "rclcpp/rclcpp.hpp"

using namespace std::chrono_literals;

struct Cmd {
  float vx{0.f};
  float vy{0.f};
  float wz{0.f};
};

class StdinRaw {
 public:
  StdinRaw() : fd_(::fileno(stdin)), active_(false) {
    if (fd_ < 0) return;
    if (::tcgetattr(fd_, &old_) != 0) return;
    termios raw = old_;
    raw.c_lflag &= static_cast<unsigned>(~(ICANON | ECHO));
    raw.c_cc[VMIN] = 0;
    raw.c_cc[VTIME] = 0;
    if (::tcsetattr(fd_, TCSANOW, &raw) == 0) active_ = true;
  }

  ~StdinRaw() {
    if (active_) {
      (void)::tcsetattr(fd_, TCSANOW, &old_);
    }
  }

 private:
  int fd_;
  termios old_;
  bool active_;
};

class CmdVelRepeater : public rclcpp::Node {
 public:
  CmdVelRepeater() : Node("phonebot_cmd_vel_repeater_cpp") {
    this->declare_parameter<std::string>("mode", "joystick");  // joystick | keyboard
    this->declare_parameter<double>("publish_hz", 50.0);
    this->declare_parameter<std::string>("cmd_vel_in_topic", "/cmd_vel");
    this->declare_parameter<std::string>("cmd_vel_out_topic", "/phonebot/cmd_vel");
    this->declare_parameter<double>("vx_max", 0.5);
    this->declare_parameter<double>("vy_max", 0.25);
    this->declare_parameter<double>("wz_max", 1.0);

    mode_ = this->get_parameter("mode").as_string();
    hz_ = this->get_parameter("publish_hz").as_double();
    in_topic_ = this->get_parameter("cmd_vel_in_topic").as_string();
    out_topic_ = this->get_parameter("cmd_vel_out_topic").as_string();
    vx_max_ = static_cast<float>(this->get_parameter("vx_max").as_double());
    vy_max_ = static_cast<float>(this->get_parameter("vy_max").as_double());
    wz_max_ = static_cast<float>(this->get_parameter("wz_max").as_double());

    pub_ = this->create_publisher<geometry_msgs::msg::Twist>(out_topic_, 10);

    if (mode_ == "joystick") {
      sub_ = this->create_subscription<geometry_msgs::msg::Twist>(
          in_topic_, 10, std::bind(&CmdVelRepeater::on_cmd_vel, this, std::placeholders::_1));
      RCLCPP_INFO(this->get_logger(), "Mode=joystick, repeating %s -> %s @ %.1f Hz", in_topic_.c_str(),
                  out_topic_.c_str(), hz_);
    } else if (mode_ == "keyboard") {
      RCLCPP_INFO(this->get_logger(), "Mode=keyboard, publishing %s @ %.1f Hz", out_topic_.c_str(), hz_);
      RCLCPP_INFO(this->get_logger(), "Keyboard: arrows for vx/vy, a/d for wz, Shift gives max. Space stops.");
      start_keyboard();
    } else {
      throw std::runtime_error("Unknown mode: " + mode_);
    }

    const auto period = std::chrono::duration<double>(1.0 / std::max(hz_, 1e-3));
    timer_ = this->create_wall_timer(
        std::chrono::duration_cast<std::chrono::nanoseconds>(period), std::bind(&CmdVelRepeater::on_timer, this));
  }

  ~CmdVelRepeater() override { stop_keyboard(); }

 private:
  void on_cmd_vel(const geometry_msgs::msg::Twist::SharedPtr msg) {
    std::lock_guard<std::mutex> lg(m_);
    latest_.vx = static_cast<float>(msg->linear.x);
    latest_.vy = static_cast<float>(msg->linear.y);
    latest_.wz = static_cast<float>(msg->angular.z);
    latest_rx_sec_ = now_sec();
  }

  void on_timer() {
    if (mode_ == "keyboard") {
      const double now = now_sec();
      std::lock_guard<std::mutex> lg(m_);
      if (kbd_vx_ts_ > 0.0 && (now - kbd_vx_ts_) > kbd_timeout_s_) latest_.vx = 0.f;
      if (kbd_vy_ts_ > 0.0 && (now - kbd_vy_ts_) > kbd_timeout_s_) latest_.vy = 0.f;
      if (kbd_wz_ts_ > 0.0 && (now - kbd_wz_ts_) > kbd_timeout_s_) latest_.wz = 0.f;
    }

    Cmd cmd;
    {
      std::lock_guard<std::mutex> lg(m_);
      cmd = latest_;
    }
    geometry_msgs::msg::Twist out;
    out.linear.x = cmd.vx;
    out.linear.y = cmd.vy;
    out.angular.z = cmd.wz;
    pub_->publish(out);
  }

  void start_keyboard() {
    kbd_stop_ = false;
    kbd_thread_ = std::thread([this]() { keyboard_loop(); });
  }

  void stop_keyboard() {
    kbd_stop_ = true;
    if (kbd_thread_.joinable()) {
      kbd_thread_.join();
    }
  }

  static double now_sec() {
    const auto ns = std::chrono::steady_clock::now().time_since_epoch();
    return std::chrono::duration<double>(ns).count();
  }

  void apply_cmd(float vx, float vy, float wz) {
    std::lock_guard<std::mutex> lg(m_);
    latest_.vx = vx;
    latest_.vy = vy;
    latest_.wz = wz;
    const double now = now_sec();
    kbd_vx_ts_ = now;
    kbd_vy_ts_ = now;
    kbd_wz_ts_ = now;
  }

  void set_vx(float vx) {
    std::lock_guard<std::mutex> lg(m_);
    latest_.vx = vx;
    kbd_vx_ts_ = now_sec();
  }

  void set_vy(float vy) {
    std::lock_guard<std::mutex> lg(m_);
    latest_.vy = vy;
    kbd_vy_ts_ = now_sec();
  }

  void set_wz(float wz) {
    std::lock_guard<std::mutex> lg(m_);
    latest_.wz = wz;
    kbd_wz_ts_ = now_sec();
  }

  void keyboard_loop() {
    StdinRaw raw;
    std::string buf;
    buf.reserve(64);

    while (rclcpp::ok() && !kbd_stop_) {
      char tmp[32];
      const ssize_t n = ::read(::fileno(stdin), tmp, sizeof(tmp));
      if (n <= 0) {
        std::this_thread::sleep_for(50ms);
        continue;
      }
      buf.append(tmp, tmp + n);

      while (!buf.empty()) {
        const unsigned char c0 = static_cast<unsigned char>(buf[0]);
        if (c0 == ' ') {
          apply_cmd(0.f, 0.f, 0.f);
          buf.erase(0, 1);
          continue;
        }
        if (c0 == 'q' || c0 == 0x03) {  // q or Ctrl+C
          rclcpp::shutdown();
          return;
        }
        if (c0 == 'a' || c0 == 'A' || c0 == 'd' || c0 == 'D') {
          const bool full = (c0 == 'A' || c0 == 'D');
          const float scale = full ? 1.0f : 0.5f;
          const float sign = (c0 == 'a' || c0 == 'A') ? 1.0f : -1.0f;
          set_wz(sign * scale * wz_max_);
          buf.erase(0, 1);
          continue;
        }

        if (c0 != 0x1b) {  // ESC
          buf.erase(0, 1);
          continue;
        }

        // Need enough bytes for parsing.
        if (buf.size() < 3) break;

        auto starts = [&](const char* s) -> bool { return buf.rfind(s, 0) == 0; };

        // Standard arrows: ESC [ A/B/C/D
        if (starts("\x1b[A")) {
          set_vx(0.5f * vx_max_);
          buf.erase(0, 3);
          continue;
        }
        if (starts("\x1b[B")) {
          set_vx(-0.5f * vx_max_);
          buf.erase(0, 3);
          continue;
        }
        if (starts("\x1b[D")) {
          set_vy(0.5f * vy_max_);  // left -> +vy
          buf.erase(0, 3);
          continue;
        }
        if (starts("\x1b[C")) {
          set_vy(-0.5f * vy_max_);  // right -> -vy
          buf.erase(0, 3);
          continue;
        }

        // Common shifted arrows: ESC [ 1 ; 2 A/B/C/D
        if (starts("\x1b[1;2A")) {
          set_vx(1.0f * vx_max_);
          buf.erase(0, std::strlen("\x1b[1;2A"));
          continue;
        }
        if (starts("\x1b[1;2B")) {
          set_vx(-1.0f * vx_max_);
          buf.erase(0, std::strlen("\x1b[1;2B"));
          continue;
        }
        if (starts("\x1b[1;2D")) {
          set_vy(1.0f * vy_max_);
          buf.erase(0, std::strlen("\x1b[1;2D"));
          continue;
        }
        if (starts("\x1b[1;2C")) {
          set_vy(-1.0f * vy_max_);
          buf.erase(0, std::strlen("\x1b[1;2C"));
          continue;
        }

        // Unrecognized escape; drop one byte and retry.
        buf.erase(0, 1);
      }
    }
  }

  std::string mode_;
  double hz_{50.0};
  std::string in_topic_;
  std::string out_topic_;
  float vx_max_{0.5f};
  float vy_max_{0.25f};
  float wz_max_{1.0f};

  rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr pub_;
  rclcpp::Subscription<geometry_msgs::msg::Twist>::SharedPtr sub_;
  rclcpp::TimerBase::SharedPtr timer_;

  std::mutex m_;
  Cmd latest_{};
  double latest_rx_sec_{0.0};

  std::thread kbd_thread_;
  bool kbd_stop_{false};
  double kbd_vx_ts_{0.0};
  double kbd_vy_ts_{0.0};
  double kbd_wz_ts_{0.0};
  double kbd_timeout_s_{0.35};
};

int main(int argc, char** argv) {
  rclcpp::init(argc, argv);
  try {
    auto node = std::make_shared<CmdVelRepeater>();
    rclcpp::spin(node);
  } catch (const std::exception& e) {
    std::fprintf(stderr, "cmd_vel_repeater exception: %s\n", e.what());
  }
  if (rclcpp::ok()) rclcpp::shutdown();
  return 0;
}

File: Ros2_bridge/src/joystick_python/joystick_python/cmd_vel_repeater.py
import select
import sys
import termios
import threading
import time
import tty
from dataclasses import dataclass
from typing import Optional, Tuple

import rclpy
from geometry_msgs.msg import Twist
from rclpy.node import Node


@dataclass
class Cmd:
    vx: float = 0.0
    vy: float = 0.0
    wz: float = 0.0


def _clamp(x: float, lo: float, hi: float) -> float:
    return min(max(x, lo), hi)


def _make_twist(cmd: Cmd) -> Twist:
    m = Twist()
    m.linear.x = float(cmd.vx)
    m.linear.y = float(cmd.vy)
    m.angular.z = float(cmd.wz)
    return m


class _StdinRaw:
    def __init__(self) -> None:
        self._fd = sys.stdin.fileno()
        self._old: Optional[Tuple] = None

    def __enter__(self) -> "_StdinRaw":
        self._old = termios.tcgetattr(self._fd)
        tty.setcbreak(self._fd)
        return self

    def __exit__(self, exc_type, exc, tb) -> None:
        if self._old is not None:
            termios.tcsetattr(self._fd, termios.TCSADRAIN, self._old)


class CmdVelRepeater(Node):
    """
    Publishes `/phonebot/cmd_vel` at a constant rate (default 50Hz).

    Modes:
    - joystick: subscribe `/cmd_vel` (Twist), repeat last command at fixed rate
    - keyboard: read keyboard from stdin, generate command, publish at fixed rate
    """

    def __init__(self) -> None:
        super().__init__("phonebot_cmd_vel_repeater_py")

        self.declare_parameter("mode", "joystick")  # joystick | keyboard
        self.declare_parameter("publish_hz", 50.0)
        self.declare_parameter("cmd_vel_in_topic", "/cmd_vel")
        self.declare_parameter("cmd_vel_out_topic", "/phonebot/cmd_vel")

        # Defaults match `teleop_twist_joy` params in `joystick_config.yaml`.
        self.declare_parameter("vx_max", 0.5)
        self.declare_parameter("vy_max", 0.25)
        self.declare_parameter("wz_max", 1.0)

        self._mode = str(self.get_parameter("mode").value)
        self._hz = float(self.get_parameter("publish_hz").value)
        self._in_topic = str(self.get_parameter("cmd_vel_in_topic").value)
        self._out_topic = str(self.get_parameter("cmd_vel_out_topic").value)
        self._vx_max = float(self.get_parameter("vx_max").value)
        self._vy_max = float(self.get_parameter("vy_max").value)
        self._wz_max = float(self.get_parameter("wz_max").value)

        self._lock = threading.Lock()
        self._latest = Cmd()
        self._latest_rx = 0.0

        self._kbd_thread: Optional[threading.Thread] = None
        self._kbd_stop = threading.Event()
        self._kbd_vx_ts = 0.0
        self._kbd_vy_ts = 0.0
        self._kbd_wz_ts = 0.0
        self._kbd_timeout_s = 0.35

        if self._mode == "joystick":
            self._sub = self.create_subscription(Twist, self._in_topic, self._on_cmd_vel, 10)
            self.get_logger().info(f"Mode=joystick, repeating {self._in_topic} -> {self._out_topic} @ {self._hz} Hz")
        elif self._mode == "keyboard":
            self._sub = None
            self.get_logger().info(f"Mode=keyboard, publishing {self._out_topic} @ {self._hz} Hz")
            self.get_logger().info("Keyboard: arrows for vx/vy, a/d for wz, Shift gives max. Space stops.")
            self._start_keyboard()
        else:
            raise RuntimeError(f"Unknown mode: {self._mode}")

        self._pub = self.create_publisher(Twist, self._out_topic, 10)
        period = 1.0 / max(self._hz, 1e-3)
        self._timer = self.create_timer(period, self._on_timer)

    def destroy_node(self) -> bool:
        self._stop_keyboard()
        return super().destroy_node()

    def _on_cmd_vel(self, msg: Twist) -> None:
        with self._lock:
            self._latest = Cmd(vx=float(msg.linear.x), vy=float(msg.linear.y), wz=float(msg.angular.z))
            self._latest_rx = time.time()

    def _on_timer(self) -> None:
        if self._mode == "keyboard":
            # Per-axis timeout so vx/vy/wz can combine (diagonal), but each axis returns to 0
            # if it hasn't been updated recently.
            now = time.time()
            with self._lock:
                if self._kbd_vx_ts > 0.0 and (now - self._kbd_vx_ts) > self._kbd_timeout_s:
                    self._latest.vx = 0.0
                if self._kbd_vy_ts > 0.0 and (now - self._kbd_vy_ts) > self._kbd_timeout_s:
                    self._latest.vy = 0.0
                if self._kbd_wz_ts > 0.0 and (now - self._kbd_wz_ts) > self._kbd_timeout_s:
                    self._latest.wz = 0.0

        with self._lock:
            cmd = self._latest

        self._pub.publish(_make_twist(cmd))

    def _start_keyboard(self) -> None:
        self._kbd_stop.clear()
        self._kbd_thread = threading.Thread(target=self._keyboard_loop, name="phonebot-kbd", daemon=True)
        self._kbd_thread.start()

    def _stop_keyboard(self) -> None:
        self._kbd_stop.set()
        t = self._kbd_thread
        if t is not None and t.is_alive():
            t.join(timeout=0.5)
        self._kbd_thread = None

    def _keyboard_loop(self) -> None:
        buf = b""

        def set_vx(vx: float) -> None:
            with self._lock:
                self._latest.vx = float(vx)
                self._kbd_vx_ts = time.time()

        def set_vy(vy: float) -> None:
            with self._lock:
                self._latest.vy = float(vy)
                self._kbd_vy_ts = time.time()

        def set_wz(wz: float) -> None:
            with self._lock:
                self._latest.wz = float(wz)
                self._kbd_wz_ts = time.time()

        def stop_all() -> None:
            with self._lock:
                self._latest = Cmd()
                now = time.time()
                self._kbd_vx_ts = now
                self._kbd_vy_ts = now
                self._kbd_wz_ts = now

        with _StdinRaw():
            while rclpy.ok() and not self._kbd_stop.is_set():
                r, _, _ = select.select([sys.stdin], [], [], 0.05)
                if not r:
                    continue
                chunk = sys.stdin.buffer.read1(32)
                if not chunk:
                    continue
                buf += chunk

                # Handle single-char commands.
                while buf:
                    b0 = buf[:1]
                    if b0 == b" ":
                        stop_all()
                        buf = buf[1:]
                        continue
                    if b0 in (b"q", b"\x03"):  # q or Ctrl+C
                        self.get_logger().info("Keyboard quit requested.")
                        rclpy.shutdown()
                        return
                    if b0 in (b"a", b"A", b"d", b"D"):
                        half = 0.5
                        full = 1.0
                        is_full = (b0 in (b"A", b"D"))
                        scale = full if is_full else half
                        wz = (scale * self._wz_max) * (1.0 if b0 in (b"a", b"A") else -1.0)
                        set_wz(wz)
                        buf = buf[1:]
                        continue

                    # Escape sequences (arrows)
                    if b0 != b"\x1b":
                        buf = buf[1:]
                        continue

                    # Need more bytes.
                    if len(buf) < 3:
                        break

                    # Standard arrows: ESC [ A/B/C/D
                    if buf.startswith(b"\x1b[A"):
                        set_vx(0.5 * self._vx_max)
                        buf = buf[3:]
                        continue
                    if buf.startswith(b"\x1b[B"):
                        set_vx(-0.5 * self._vx_max)
                        buf = buf[3:]
                        continue
                    if buf.startswith(b"\x1b[D"):
                        set_vy(0.5 * self._vy_max)  # left -> +vy
                        buf = buf[3:]
                        continue
                    if buf.startswith(b"\x1b[C"):
                        set_vy(-0.5 * self._vy_max)  # right -> -vy
                        buf = buf[3:]
                        continue

                    # Common shifted arrows in terminals: ESC [ 1 ; 2 A/B/C/D
                    if buf.startswith(b"\x1b[1;2A"):
                        set_vx(1.0 * self._vx_max)
                        buf = buf[len(b"\x1b[1;2A") :]
                        continue
                    if buf.startswith(b"\x1b[1;2B"):
                        set_vx(-1.0 * self._vx_max)
                        buf = buf[len(b"\x1b[1;2B") :]
                        continue
                    if buf.startswith(b"\x1b[1;2D"):
                        set_vy(1.0 * self._vy_max)
                        buf = buf[len(b"\x1b[1;2D") :]
                        continue
                    if buf.startswith(b"\x1b[1;2C"):
                        set_vy(-1.0 * self._vy_max)
                        buf = buf[len(b"\x1b[1;2C") :]
                        continue

                    # Unrecognized escape; drop one byte and retry.
                    buf = buf[1:]


def main() -> None:
    rclpy.init()
    node = None
    try:
        node = CmdVelRepeater()
        rclpy.spin(node)
    finally:
        if node is not None:
            node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == "__main__":
    main()

